Addendum to Arcade build

mstv opened the pull request:
Addendum to #7649, solves parts of #7702.

## Proposed changes

- Fixup merge of `UnitTests/GitCommands.Tests/AppTitleGeneratorTests.cs`
- Set `NUnitConsoleRunnerVersion` in `DefaultVersions.props`
- Prepare to run unit tests in parallel, keep integation tests serial but
- Do not run any tests in parallel until stable
- Prepare to repeat test runs
- Store test results as artifacts
- Ignore test fails if the AppVeyor environment variable `IGNORE_TEST_RESULT` is `true`.

## Test methodology <!-- How did you ensure quality? -->

- run the NUnit tests

## Test environment(s) <!-- Remove any that don't apply -->

- GIT <!-- Add version 2.11 or above -->
- Windows <!-- Add version 7 SP1 or above -->

<!-- Mention language, UI scaling, or anything else that might be relevant -->

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
Should be fixed now.

mstv commented:
I cannot rebuild on AppVeyor also after successful sign out/in:
`Error reading PR #7703 details from GitHub: GitHub authorization required.`
![grafik](https://user-images.githubusercontent.com/36601201/73472056-bb96ee00-438a-11ea-91c5-da6681c50271.png)

gerhardol commented on UnitTests/GitCommands.Tests/AppTitleGeneratorTests.cs:
```diff
@@ -62,7 +62,7 @@ public void Generate_should_include_debug_suffix([Values(null, "invalid")] strin
             var title = _appTitleGenerator.Generate("a", true, null);
 
-            title.Should().Be($"{ShortName} (no branch) - Git Extensions");
+            title.Should().Be($"{ShortName} (no branch) - Git Extensions [DEBUG]");
```
OT
The test on line 76 will fail if the sha length is made "dynamic".

mstv commented:
How to view `build.binlog`?

Local test execution fails without meaningful error message:
```
f:\Build\gitextensions3>cibuild.cmd -integrationTest                                                                    task.setvariable isSecret=False;variable=Artifacts;isOutput=TrueF:\Build\gitextensions3\artifacts
task.setvariable isSecret=False;variable=Artifacts.Toolset;isOutput=TrueF:\Build\gitextensions3\artifacts\toolset
task.setvariable isSecret=False;variable=Artifacts.Log;isOutput=TrueF:\Build\gitextensions3\artifacts\log\Release
task.setvariable isSecret=False;variable=TEMP;isOutput=TrueF:\Build\gitextensions3\artifacts\tmp\Release
task.setvariable isSecret=False;variable=TMP;isOutput=TrueF:\Build\gitextensions3\artifacts\tmp\Release
Downloading vswhere
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\msbuild.exe /nologo /bl:F:\Build\gitextensions3\artifacts\log\Release\build.binlog /clp:Summary /m /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:Configuration=Release /p:RepoRoot=F:\Build\gitextensions3 /p:Restore=False /p:Build=False /p:Rebuild=False /p:Test=False /p:IntegrationTest=True /v:minimal /warnaserror F:\Build\gitextensions3\Build\tools\Build.proj
  Running tests: F:\Build\gitextensions3\artifacts\bin\tests\IntegrationTests\UI.IntegrationTests\net461\UI.Integration
  Tests.dll [net461|x64]
NUnit : error :  [net461|x64] [F:\Build\gitextensions3\IntegrationTests\UI.IntegrationTests\UI.IntegrationTests.csproj]

Fehler beim Buildvorgang.

NUnit : error :  [net461|x64] [F:\Build\gitextensions3\IntegrationTests\UI.IntegrationTests\UI.IntegrationTests.csproj]
    0 Warnung(en)
    1 Fehler

Verstrichene Zeit 00:00:05.00
Microsoft.PowerShell.Commands.WriteErrorException: Build failed.
-1

f:\Build\gitextensions3>
```

mstv commented:
`cibuild.cmd` failed due to missing `nunit3-console.exe`.
`build.cmd` works better.
But several tests which passed before, often fail now, e.g.
```
1) Failed : GitUITests.CancellationTokenSequenceTests.Concurrent_callers_to_Next_only_result_in_one_non_cancelled_token_being_issued
  Test should have completed within a reasonable amount of time
  Expected: True
  But was:  False
   bei GitUITests.CancellationTokenSequenceTests.<Concurrent_callers_to_Next_only_result_in_one_non_cancelled_token_being_issued>d__8.MoveNext() in F:\Build\gitextensions3\UnitTests\GitUI.Tests\CancellationTokenSequenceTests.cs:Zeile 135.
   bei System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
   bei GitUITests.CancellationTokenSequenceTests.Concurrent_callers_to_Next_only_result_in_one_non_cancelled_token_being_issued()
```

mstv commented:
AppVeyor builds are not triggered by commits anymore.

mstv commented:
Which version of `nunit3-console.exe` / `NUnit.ConsoleRunner` is required, @sharwell?
Currently we use:
```
            "NUnit": {
              "suppressParent": "All",
              "target": "Package",
              "version": "[3.12.0, )",
              "autoReferenced": true
            },
            "NUnit.ConsoleRunner": {
              "suppressParent": "All",
              "target": "Package",
              "version": "[3.10.0, )",
              "autoReferenced": true
            },
            "NUnit3TestAdapter": {
              "suppressParent": "All",
              "target": "Package",
              "version": "[3.16.1, )",
              "autoReferenced": true
            },
```

mstv commented:
> Which version...?

OK. NuGet does not list newer versions.

RussKie commented:
> `cibuild.cmd` failed due to missing `nunit3-console.exe`.

If you want to run it locally you **must** have `nunit3-console.exe` available on PATH. AppVeyor uses a version with custom plugins that facilitate reporting of tests to [Test] tab.

`cibuild.cmd` is made to accommodate CI systems and it is not designed to be used by developers.

RussKie commented:
> How to view `build.binlog`? [#7702 (comment)](https://github.com/gitextensions/gitextensions/issues/7702#issuecomment-580212329)

![image](https://user-images.githubusercontent.com/4403806/73497566-31449d80-440f-11ea-81e3-c522c36e6458.png)

https://github.com/KirillOsenkov/MSBuildStructuredLog

mstv commented:
I'd like to merge this.

pmiossec commented:
I have found something not working with the arcade build.

Due to how output folders are structured, I have found that one plugin 'Azure Devops CI' is not working well while debugging.
A (silent, because plugins failed without feedback 😕) FileNotFoundException is thrown because it fails to load a dependency assembly 'NewtonSoft' even if it is in the plugin folder. 

Copying the file in Gitextensions folder solve the problem. So it seems that the dependency is loaded from the exe folder. 

Perhaps it happens for other plugins because I could see a lot of 'Filenotfoundexception' in the console.

It needs more investigation but I don't have the time at the moment...

RussKie commented:
> I have found something not working with the arcade build.

Please raise a new issue.
The new folder structure kind of mimics what the current installation structure is, so everything ought to work. If it is not, we certainly need to investigate.

RussKie commented:
> I'd like to merge this.

I think I got AppVeyor back up, so once it builds we can merge it.

RussKie commented on appveyor.yml:
```diff
@@ -59,8 +60,41 @@ test_script:
+                $total  = $result[1].Value
+                $passed = $result[4].Value
+                $failed = $result[5].Value
+                $notPassed = $total - $passed
```
This isn't used

mstv replied:
Yes, `$notPassed` would be a safe alternative to `$failed` as number of `$testErrors`. But it is too strict because we also have ignored / disabled testcases.

RussKie replied:
...so remove?

RussKie commented on appveyor.yml:
```diff
@@ -59,8 +60,41 @@ test_script:
+                Push-AppveyorArtifact $resultFilePath -FileName $artifactName
+                if ($testErrors -ne 0) {
+                    $logFile = Get-ChildItem -recurse artifacts\$testname*.log
+                    Get-Content $logFile
```
Is this to output the content of the file with failed tests? Not sure, I think it may need to be:
```suggestion
                    Get-Content $logFile | Out-Host
```

Have you tested this on AppVeyor? It has its own idiosyncrasies...

mstv replied:
Yes, it is the output of failed tests. It works on AppVeyor.

RussKie commented on appveyor.yml:
```diff
@@ -59,8 +60,41 @@ test_script:
+                $testErrors = $failed
+                $testName = $resultFilePath.Split('\\')[-3]
+                $artifactName = "TestResult#${testRun}.${testErrors}err.${testName}.xml"
+                Push-AppveyorArtifact $resultFilePath -FileName $artifactName
```
Do we need to publish these?

mstv replied:
Yes, because they contain the full truth ;-). Instead of these, I have omitted the `.log` of successful tests.

RussKie replied:
Is there a benefit in publishing them for successful tests though?

mstv replied:
We can omit them when the tests pass stably. Until then, they might contain e.g. printf debug output for comparison with failing tests.
In addition, we have a test which is not executed although it should: #7695. This is visible in the `TestResult.xml`, too.

RussKie commented on appveyor.yml:
```diff
@@ -14,6 +14,7 @@ environment:
   ARCHIVE_WITH_PDB: TRUE
   GE_TEST_SLEEP_SECONDS_ON_HANG: 0
   GE_TEST_LAUNCH_DEBUGGER_ON_HANG: 0
+  TEST_RUN_COUNT: 1
```
When you set it via AppVeyor UI does it override this value?

mstv replied:
Yes, these variables can be overridden using the AppVeyor project settings.
This makes it difficult to override them for one PR temporarily.
Then we should mark them as default values at least.

RussKie commented:
@mstv I suggest we squash this before merge

RussKie commented:
> cannot rebuild on AppVeyor

Please let me know if you still have troubles and ping on gitter or via email

mstv commented:
> I suggest we squash this before merge

It's only 6 not very related commits. Especially, https://github.com/gitextensions/gitextensions/pull/7703/commits/74645259b729164fa946b7ff77fc27c387547f34 is a candidate for reversion.

RussKie commented:
> Especially, [7464525](https://github.com/gitextensions/gitextensions/commit/74645259b729164fa946b7ff77fc27c387547f34) is a candidate for reversion.

We can always add it later.
Feel like we adding internal details to the main trunk

mstv commented:
> Feel like we adding internal details to the main trunk

They will be visble anyway -- just more or less easy to comprehend. It doesn't greatly matter for me.

mstv commented:
I have rebased and squashed partially. I am going to merge this tomorrow.
This PR lets fail the AppVeyor builds on test fail again. The test fails can be ignored by setting the AppVeyor environment variable `IGNORE_TEST_RESULT` to `true`.
As the test fail very often currently, I have activated this in the AppVeyor settings.

RussKie commented:
👍

On Thu, Feb 6, 2020, 6:59 AM Michael Seibt <notifications@github.com> wrote:

> I have rebased and squashed partially. I am going to merge this tomorrow.
> This PR lets fail the AppVeyor builds on test fail again. The test fails
> can be ignored by setting the AppVeyor environment variable
> IGNORE_TEST_RESULT to true.
> As the test fail very often currently, I have activated this in the
> AppVeyor settings.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7703?email_source=notifications&email_token=ABBTEXXEJNOMAMH2RSZGMJ3RBMLBLA5CNFSM4KNMMH7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEK4YVRY#issuecomment-582585031>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTSY2HTSMDDHK3XIODRBMLBLANCNFSM4KNMMH7A>
> .
>

mstv commented:
@msftbot merge in 1 day
